include/revng/Model/Type.h

```diff
@@ -0,0 +1,726 @@
+    io.enumCase(Val, "Pointer", model::QualifierKind::Pointer);
+    io.enumCase(Val, "Array", model::QualifierKind::Array);
+    io.enumCase(Val, "Const", model::QualifierKind::Const);
+  }
+};
+
+namespace model {
+
+struct Qualifier;
+struct QualifiedType;
+
+struct EnumEntry;
+struct AggregateField;
+struct StructField;
+struct UnionField;
+struct ArgumentType;
+
+} // end namespace model
+
+// Make model::Type derived types usable with UpcastablePointer
+template<>
+struct concrete_types_traits<model::Type> {
+  using type = std::tuple<model::PrimitiveType,
+                          model::EnumType,
+                          model::TypedefType,
+                          model::StructType,
+                          model::UnionType,
+                          model::FunctionPointerType>;
+};
+
+template<>
+struct concrete_types_traits<const model::Type> {
+  using type = std::tuple<const model::PrimitiveType,
+                          const model::EnumType,
+                          const model::TypedefType,
+                          const model::StructType,
+                          const model::UnionType,
+                          const model::FunctionPointerType>;
+};
+
+/// \brief Concept to identify all types that are derived from model::Type
+template<typename T>
+concept FlatCType = std::is_base_of_v<model::Type, T>;
+
+/// \brief Make all types derived from model::Type yaml-serializable like tuples
+template<FlatCType T>
+struct llvm::yaml::MappingTraits<T> : public TupleLikeMappingTraits<T> {};
+
+/// \brief Base class of model types used for LLVM-style RTTI
+struct model::Type {
+  // static constexpr const char *Tag = "!Invalid";
+  using Key = std::pair<model::TypeKind::Values, uint64_t>;
+
+  /// \brief The kind of a model::Type
+  TypeKind::Values Kind;
+
+  // TODO: we should switch to a sane UUID implementation
+  uint64_t ID;
+
+  /// \brief The name of this type. This is the unique key to identifiy a type.
+  std::string Name;
+
+public:
+  static bool classof(const Type *T) { return false; }
+
+  Key key() const { return Key{ Kind, ID }; }
+
+protected:
+  Type(TypeKind::Values TK, llvm::StringRef NameRef) : Type(TK, 0ULL, NameRef) {
+    auto Err = llvm::getRandomBytes(&ID, sizeof(ID));
+    if (Err)
+      ID = 0ULL;
+  }
+
+  Type(TypeKind::Values TK, uint64_t TheID, llvm::StringRef NameRef) :
+    Kind(TK), ID(TheID), Name(NameRef) {}
+
+public:
+  template<FlatCType T>
+  bool operator==(const T &Other) const {
+    return key() == Other.key();
+  }
+
+  template<FlatCType T>
+  bool operator<=(const T &Other) const {
+    return key() <= Other.key();
+  }
+
+protected:
+  bool verify() const debug_function;
+};
+INTROSPECTION_NS(model, Type, Kind, ID, Name);
+
+namespace model {
+
+using UpcastableType = UpcastablePointer<model::Type>;
+
+} // end namespace model
+
+template<>
+struct llvm::yaml::ScalarTraits<model::Type::Key>
+  : CompositeScalar<model::Type::Key, '#'> {};
```

Comment:
Elsewhere we use `-` as a separator. Reasons to deviate?

Reply:
Yes but nothing major. The idea is that if we ever switch to GUID in future we might end up with the standard-conforming GUID format that contains `-`. So I went for `#`. But if you prefer consistency I'll switch to `-`. In the end, if we switch to real GUID, and we have `-` in the middle, the tests should fail and we can easily replace `-` with `#` or something else at that point. Just let me know what you think and I'll do it.

Reply:
Ah, I see. For now let's stick with `-`. Switching to GUID will be a breaking change anyway.

Reply:
Done.